packages/form-js-playground/src/components/PlaygroundRoot.js

```diff
@@ -193,59 +165,77 @@ export function PlaygroundRoot(props) {
-      form.destroy();
+      inputDataEditor.destroy();
+      outputDataEditor.destroy();
+      formViewer.destroy();
       formEditor.destroy();
     };
-  }, []);
+  }, [ additionalModules, editorAdditionalModules, editorProperties, emit, exporterConfig, propertiesPanelConfig, viewerAdditionalModules, viewerProperties ]);
 
+  // initialize data through props
   useEffect(() => {
-    dataEditorRef.current.setValue(toString(initialData));
-  }, [ initialData ]);
+    if (!config.initialSchema) {
+      return;
+    }
+
+    load(config.initialSchema, config.initialData || {});
+  }, [ config.initialSchema, config.initialData, load ]);
 
   useEffect(() => {
-    if (initialSchema) {
-      formEditorRef.current.importSchema(initialSchema);
-      dataEditorRef.current.setVariables(getSchemaVariables(initialSchema));
-    }
-  }, [ initialSchema ]);
+    schema && formViewerRef.current.importSchema(schema, data);
+  }, [ schema, data ]);
 
   useEffect(() => {
-    if (schema && dataContainerRef.current) {
+    if (schema && inputDataContainerRef.current) {
       const variables = getSchemaVariables(schema);
-      dataEditorRef.current.setVariables(variables);
+      inputDataRef.current.setVariables(variables);
     }
   }, [ schema ]);
 
+  // exposes api to parent
   useEffect(() => {
-    schema && formRef.current.importSchema(schema, data);
-  }, [ schema, data ]);
 
-  useEffect(() => {
-    resultViewRef.current.setValue(toString(resultData));
-  }, [ resultData ]);
+    if (!apiLinkTarget) {
+      return;
+    }
+
+    apiLinkTarget.api = {
+      attachDataContainer: (node) => inputDataRef.current.attachTo(node),
+      attachResultContainer: (node) => outputDataRef.current.attachTo(node),
+      attachFormContainer: (node) => formViewerRef.current.attachTo(node),
+      attachEditorContainer: (node) => formEditorRef.current.attachTo(node),
+      attachPaletteContainer: (node) => formEditorRef.current.get('palette').attachTo(node),
+      attachPropertiesPanelContainer: (node) => formEditorRef.current.get('propertiesPanel').attachTo(node),
+      get: (name, strict) => formEditorRef.current.get(name, strict),
+      getDataEditor: () => inputDataRef.current,
+      getEditor: () => formEditorRef.current,
+      getForm: () => formViewerRef.current,
+      getResultView: () => outputDataRef.current,
+      getSchema: () => formEditorRef.current.getSchema(),
+      saveSchema: () => formEditorRef.current.saveSchema(),
+      setSchema: setSchema,
+      setData: setData
+    };
 
+  }, [ apiLinkTarget ]);
+
+  // separate effect for state to avoid re-creating the api object every time
   useEffect(() => {
-    onStateChanged && onStateChanged({
-      schema,
-      data
-    });
-  }, [ onStateChanged, schema, data ]);
 
-  const handleDownload = useCallback(() => {
+    if (!apiLinkTarget) {
+      return;
+    }
+
+    apiLinkTarget.api.getState = () => ({ schema, data });
+    apiLinkTarget.api.load = load;
```

Comment:
So we have two ways to load data: through the initial props and this api method.

Reply:
Yes exactly. I agree this isn't great and not react-like, but I didn't want to make any changes to any of the logic.